### DISCUSSION TITLE
How to mesh 2 phase microstructure

### DISCUSSION THREAD

**Discussion by smitra43** [idaholab/moose#27259] [Q&A Meshing]
I have a large number of square images. they are all binary images of 2 of the same phases in different blends. I want to simulate elastic deformation in them. my intention is to do a uniform mesh where each pixel is an element of my mesh. but how can I input an image (or numpy array for image) into moose. how will it assign different stiffness values depending on the location? Here is my code:

```
import numpy as np
import h5py
import os

def read_microstructure_from_hdf5(filepath, dataset_name="dataset"):
    with h5py.File(filepath, 'r') as file:
        microstructure_data = np.array(file[dataset_name][168])
    return microstructure_data

def generate_moose_input_file(input_file_path, image_dim):
    with open(input_file_path, 'w') as file:
        file.writelines([
            "[GlobalParams]\n",
            "  order = FIRST\n",
            "  family = LAGRANGE\n",
            "  displacements = 'disp_x disp_y'\n",
            "  out_of_plane_strain = strain_zz\n",
            "[]\n\n",
            
            "[Mesh]\n",
            "  [./square]\n",
            "    type = GeneratedMeshGenerator\n",
            f"    dim = 2\n",
            f"    nx = {image_dim}\n",
            f"    ny = {image_dim}\n",
            "    elem_type = QUAD4\n",
            "  [../]\n",
            "[]\n\n",
            
            "[Variables]\n",
            "  [./disp_x]\n",
            "  [../]\n",
            "  [./disp_y]\n",
            "  [../]\n",
            "  [./strain_zz]\n",
            "  [../]\n",
            "[]\n\n",
            
            "[AuxVariables]\n",
            "  [./nl_strain_zz]\n",
            "    order = CONSTANT\n",
            "    family = MONOMIAL\n",
            "  [../]\n",
            "[]\n\n",
            
            "[Materials]\n",
            "  [./phase1]\n",
            "    type = GenericConstantMaterial\n",
            "    properties = 'E nu'\n",
            "    property_values = '500 0.2'\n",
            "  [../]\n",
            "  [./phase2]\n",
            "    type = GenericConstantMaterial\n",
            "    properties = 'E nu'\n",
            "    property_values = '6000 0.3'\n",
            "  [../]\n",
            "[]\n\n",
            
            "[BCs]\n",
            "  [./left]\n",
            "    type = DirichletBC\n",
            "    variable = disp_x\n",
            "    boundary = left\n",
            "    value = 0.0\n",
            "  [../]\n",
            "  [./bottom]\n",
            "    type = DirichletBC\n",
            "    variable = disp_y\n",
            "    boundary = bottom\n",
            "    value = 0.0\n",
            "  [../]\n",
            "[]\n\n",
            
            "[Executioner]\n",
            "  type = Steady\n",
            "  solve_type = 'NEWTON'\n",
            "  petsc_options_iname = '-pc_type -pc_hypre_type'\n",
            "  petsc_options_value = 'hypre boomeramg'\n",
            "  nl_rel_tol = 1e-5\n",
            "  nl_abs_tol = 1e-8\n",
            "  nl_max_its = 50\n",
            "[]\n\n",
            
            "[Outputs]\n",
            "  exodus = true\n",
            "[]\n\n",
        ])

def process_image_to_input_file(filepath, output_folder, dataset_name="dataset"):
    binary_image = read_microstructure_from_hdf5(filepath, dataset_name)
    image_dim = binary_image.shape[0]
    base_name = os.path.splitext(os.path.basename(filepath))[0]
    input_file_path = os.path.join(output_folder, f"{base_name}.i")

    generate_moose_input_file(input_file_path, image_dim)

def process_images_folder(folder_path, output_folder, dataset_name="dataset"):
    for file in os.listdir(folder_path):
        if not file.endswith('.h5') and not file.endswith('.hdf5'):
            continue
        filepath = os.path.join(folder_path, file)
        process_image_to_input_file(filepath, output_folder, dataset_name)

folder_path = ''
output_folder = ''
dataset_name = ""

process_images_folder(folder_path, output_folder, dataset_name)
```

**Comment by GiudGiud** [comment]
Hello

you can use the PropertyReadFile user object to read a CSV of the values of a space dependent field. The CSV will have to be created upstream of moose

then a PiecewiseConstantFromCSV will be able to give the spatial values of these quantities anywhere in the mesh

**Reply by GiudGiud** [reply]
So you are already creating the input through a python script here.
Processing the data into a CSV that gets loaded by the python-created script will also need to be done by a script.

**Reply by smitra43** [reply]
i have a very large dataset of images, how should I go about creating the input file besides a python script?

**Reply by GiudGiud** [reply]
Scripts are the way to go. Updated the text

**Reply by smitra43** [reply]
[MICRO2D_homogenized.txt](https://github.com/idaholab/moose/files/14831102/MICRO2D_homogenized.txt)
[MICRO2D_homogenized_material_properties.csv](https://github.com/idaholab/moose/files/14831104/MICRO2D_homogenized_material_properties.csv)
I keep getting errors like "stress not found in block 0" and other things not found in block 0.

**Reply by GiudGiud** [reply]
Stress is a material property, which should be the defined by the appropriate Material object.
Starting from an example, you should be able to find the Material you should use. It s probably some sort of strain-stress relationship materisal

**Reply by smitra43** [reply]
so these are artificial microstructures and so I won't be able to find the "material"  - hence the csv file so that I can define the young's modulus and poisson's ratio at each element. Hence I was thinking of creating a custom material but I guess this is not the way to go? 

[UserObjects]
  [./read_properties]
    type = PropertyReadFile
    prop_file_name = '/storage/coda1/p-skalidindi7/0/shared/MOOSE/MICRO2D_homogenized_mat_prop.csv'
    read_type = 'element'
    nprop = 4
  [../]
 
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]
 
[]
 
[Functions]
  [./youngs_modulus]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = read_properties
    read_type = 'element'
    column_number = 2
  [../]
  [./poissons_ratio]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = read_properties
    read_type = 'element'
    column_number = 3
  [../]
[]
 
 **[Materials]
   [./elastic_properties]
    type = CustomElasticMaterial
    youngs_modulus_function = youngs_modulus
    poissons_ratio_function = poissons_ratio
  [../]
[]**

**Reply by GiudGiud** [reply]
Hello

this looks fine to me for having element dependent material properties. 
you do need to be able to generate a CSV file ahead of the simulation 

Guillaume
